opt: check UDFs and UDTs by name when checking metadata dependencies

DrewKimball opened the pull request:
#### tree: distinguish UDFs and builtins that use a SQL string body

This refactor changes the meaning of the `Overload.IsUDF` field to
be true only for user-defined functions - meaning those created using
`CREATE FUNCTION`. This is contrasted with builtin functions that are
defined with a SQL string set in `Overload.Body`. Logic that cares only
about user-defined functions can continue checking `Overload.IsUDF`,
while cases that deal with the SQL body should use `Overload.HasSQLBody()`.
Note that it is insufficient to check whether `Overload.Body` is empty
because it is possible to define a UDF with an empty body.

#### opt: refactor metadata dependency tracking

This commit performs some refactoring for the way data source objects
are tracked in `opt.Metadata` in order to make future changes to UDT
and UDF dependency tracking easier. More particularly, UDTs and UDFs
will be able to re-resolve any references by name. This is necessary
in order to handle cases where changes to the search-path cause names
in the query to resolve to different objects.

#### opt: track UDT references by name in the Metadata

Previously, it was possible for invalid queries to be kept in the cache
after a schema change, since user-defined types were tracked only by OID.
This missed cases where the search path changed which object a name in
the query resolved to (or whether it resolved at all).

This patch fixes this behavior by tracking UDT references by name, similar
to what was already done for data sources. This ensures that the query
staleness check doesn't miss changes to the search path.

#### opt: track UDFs in the Metadata

This patch adds tracking for user-defined functions in `opt.Metadata`.
This ensures that the query cache will be correctly invalidated after
a schema change or search-path change that could change the query's
semantics, or cause it to error.

Fixes #95214
Fixes #93082
Fixes #93321
Fixes #96674

Release note (bug fix): Fixed a bug that could prevent a cached query from being invalidated when a UDF or UDT referenced by that query was altered or dropped, or when the schema or database of a UDF or UDT was altered or dropped.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96045)
<!-- Reviewable:end -->

DrewKimball commented:
Hold off on reviewing this for a bit - it should be possible to fix #93321 as well.

ajwerner commented on pkg/sql/sem/tree/function_name.go:
```diff
@@ -63,6 +63,9 @@ type FunctionReferenceResolver interface {
 type ResolvableFunctionReference struct {
+	// CallsiteName is the original UnresolvedName used at the function callsite.
+	// It may be unset if the function is unresolved.
+	CallsiteName *UnresolvedName
```
Something about this field name bothers me. I think your approach here is sound. Also, I wonder if we should make this field unexported and have the act of resolving which rewrites this in place to do the swapping out into the field. Then have a method which uses the unexported field if it's populated and otherwise delegates to the underlying function reference.

ajwerner reviewed (approved):
This looks good. My comment is nit-picky and more of a feeling than anything else. A lot of code in tree makes me feel uneasy with its unprincipled mutability etc, so there's nothing you could probably do that would make me feel good. 

Feel free to wave me off if you like what's there.

rharding6373 reviewed (changes requested):
This is great!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-:-NN8eoRk5Wde04fKuCdN:b9znln1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @DrewKimball)

___
*[`pkg/sql/opt/metadata.go` line 355 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NN8c2wzBJrn9Lny_scL:-NN8c2wzBJrn9Lny_scM:b-5aedgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2eaea52fe6d4f802bff5d8fadff770bd6971927/pkg/sql/opt/metadata.go#L355)):*
> ```Go
> 		}
> 	}
> 	// handleUndefined swallows "undefined" and "dropped object errors, since
> ```

nit: s/\"dropped object/\"dropped\" object
___
*[`pkg/sql/opt/metadata.go` line 434 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NN8bfWp7ku8L6rvuN69:-NN8bfWq5aN-UsoGyWWK:baduoq7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2eaea52fe6d4f802bff5d8fadff770bd6971927/pkg/sql/opt/metadata.go#L434)):*
> ```Go
> 	}
> 	if name == nil {
> 		panic(errors.AssertionFailedf("attempted to add UDF with nil name"))
> ```

Is there a way to error gracefully instead of panic here? Better to not panic if we can.
___
*[`pkg/sql/opt/metadata.go` line 439 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NN8Zm83AEtKQair5cQN:-NN8Zm83AEtKQair5cQO:blnun7c) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2eaea52fe6d4f802bff5d8fadff770bd6971927/pkg/sql/opt/metadata.go#L439)):*
> ```Go
> 		if md.userDefinedFunctions[i].overload == fun && md.userDefinedFunctions[i].name == name {
> 			// This is a duplicate.
> 			break
> ```

Should this be a return? Right now it looks like we will always add the UDF to the metadata whether we found a duplicate or not.

Is there a test case that covers duplicates?


<!-- Sent from Reviewable.io -->

mgartner reviewed (changes requested):
> Reviewed 3 of 6 files at r1, 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, and @DrewKimball)

___
*[`pkg/sql/opt/metadata.go` line 377 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NN9CfqO5f6Jt67vKKYX:-NN9CfqO5f6Jt67vKKYY:b-9xk8o6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2eaea52fe6d4f802bff5d8fadff770bd6971927/pkg/sql/opt/metadata.go#L377)):*
> ```Go
> 	for i := range md.userDefinedFunctions {
> 		dep := &md.userDefinedFunctions[i]
> 		toCheck, err := optCatalog.ResolveFunction(ctx, dep.name, &evalCtx.SessionData().SearchPath)
> ```

Can we resolve the overload by ID rather than by name? I think that'd be more robust than using `MatchOverload`.
___
*[`pkg/sql/opt/metadata.go` line 430 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NN9CzNe9z3XUzTlNDGs:-NN9CzNe9z3XUzTlNDGt:bcdgw4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2eaea52fe6d4f802bff5d8fadff770bd6971927/pkg/sql/opt/metadata.go#L430)):*
> ```Go
> // this query.
> func (md *Metadata) AddUserDefinedFunc(fun *tree.Overload, name *tree.UnresolvedName) {
> 	if !fun.IsUDF {
> ```

We now set `Overload.IsUDF` to true for some builtins that use the UDF mechanisms. This is some tech debt that I think is time to address. I think we'll need two different booleans:

* `Overload.IsUDF` can mean what it says - is the function user-defined.
* `Overload.HasBody` (or some better name you come up with) can mean that an Overload's implementation is defined in the `Body` field.

We can't rely on an empty `Body` for the second point in order to be _completely_ consistent with Postgres because UDFs with empty bodies are allowed.

Might be good to make this refactor in a separate PR/commit.
___
*[`pkg/sql/opt/metadata.go` line 437 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NN9E2R-BFZE0xhYmypz:-NN9E2R-BFZE0xhYmyq-:b-k3hru2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2eaea52fe6d4f802bff5d8fadff770bd6971927/pkg/sql/opt/metadata.go#L437)):*
> ```Go
> 	}
> 	for i := range md.userDefinedFunctions {
> 		if md.userDefinedFunctions[i].overload == fun && md.userDefinedFunctions[i].name == name {
> ```

Why do we need to check the name here?
___
*[`pkg/sql/sem/tree/function_name.go` line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NN6ENT9HLMi3njNbsZc-r2-68:-NN9FUln51E5OcGMns0r:b-rtbg5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2eaea52fe6d4f802bff5d8fadff770bd6971927/pkg/sql/sem/tree/function_name.go#L68)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Something about this field name bothers me. I think your approach here is sound. Also, I wonder if we should make this field unexported and have the act of resolving which rewrites this in place to do the swapping out into the field. Then have a method which uses the unexported field if it's populated and otherwise delegates to the underlying function reference. 
</blockquote></details>

If `CheckDependencies` resolves the function by ID/OID, then we shouldn't need this, right?


<!-- Sent from Reviewable.io -->

DrewKimball committed:
tree: distinguish UDFs and builtins that use a SQL string body

This refactor changes the meaning of the `Overload.IsUDF` field to
be true only for user-defined functions - meaning those created using
`CREATE FUNCTION`. This is contrasted with builtin functions that are
defined with a SQL string set in `Overload.Body`. Logic that cares only
about user-defined functions can continue checking `Overload.IsUDF`,
while cases that deal with the SQL body should use `Overload.HasSQLBody()`.
Note that it is insufficient to check whether `Overload.Body` is empty
because it is possible to define a UDF with an empty body.

Fixes #95214

Release note: None

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @mgartner, and @rharding6373)

___
*[`pkg/sql/opt/metadata.go` line 355 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NN8c2wzBJrn9Lny_scL:-NNDjlPTBwvI7VjbMzwF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2eaea52fe6d4f802bff5d8fadff770bd6971927/pkg/sql/opt/metadata.go#L355)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

nit: s/\"dropped object/\"dropped\" object
</blockquote></details>

Done.
___
*[`pkg/sql/opt/metadata.go` line 377 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NN9CfqO5f6Jt67vKKYX:-NNDlK6w8eScOFnR7Jnd:bhdq9ve) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2eaea52fe6d4f802bff5d8fadff770bd6971927/pkg/sql/opt/metadata.go#L377)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Can we resolve the overload by ID rather than by name? I think that'd be more robust than using `MatchOverload`.
</blockquote></details>

We have to use name in case the name now resolves to a different overload - for example, in `logic_test/udf` the re-created sequence will return `5` instead of `1` if we don't properly handle this. It also wouldn't handle the database being switched.
___
*[`pkg/sql/opt/metadata.go` line 430 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NN9CzNe9z3XUzTlNDGs:-NNDxz3K5yIRxIXu4x1c:b-sq80ef) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2eaea52fe6d4f802bff5d8fadff770bd6971927/pkg/sql/opt/metadata.go#L430)):*
I've refactored in a new commit. I'd appreciate some extra scrutiny for the cases where `IsUDF` is checked vs `HasSQLBody` - I left some as `IsUDF` because they seem to care only about whether the function was user-defined, not whether it uses the SQL string in its definition.

>We can't rely on an empty Body for the second point in order to be completely consistent with Postgres because UDFs with empty bodies are allowed.

That case is only when `IsUDF` is true, right? So it should be enough to check `IsUDF || Body != ""`. I've added a method to `Overload` to handle this. If you think it's better to add `HasBody`, I'm fine with that as well.
___
*[`pkg/sql/opt/metadata.go` line 434 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NN8bfWp7ku8L6rvuN69:-NNEPHix93AHr4K57FXL:bgxzerp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2eaea52fe6d4f802bff5d8fadff770bd6971927/pkg/sql/opt/metadata.go#L434)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Is there a way to error gracefully instead of panic here? Better to not panic if we can.
</blockquote></details>

I don't think there's a way to propagate an error without plumbing it through the optbuilder. I think the panic should be caught be the optimizer, anyway.
___
*[`pkg/sql/opt/metadata.go` line 437 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NN9E2R-BFZE0xhYmypz:-NNEOSQW7R651cNieb1Y:b-kwddcm) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2eaea52fe6d4f802bff5d8fadff770bd6971927/pkg/sql/opt/metadata.go#L437)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Why do we need to check the name here?
</blockquote></details>

We can't de-duplicate fully if a function is referred to in different ways (say, `db.schema.func` vs `func`), since that can change how a function is resolved in a different context (like if the database name is changed). I've changed the field from `name` to `reference` and added a comment to try and clarify this a bit. Also expanded the tests.
___
*[`pkg/sql/opt/metadata.go` line 439 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NN8Zm83AEtKQair5cQN:-NNEMdeQ-qgwi2RZI33C:bss28k3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2eaea52fe6d4f802bff5d8fadff770bd6971927/pkg/sql/opt/metadata.go#L439)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Should this be a return? Right now it looks like we will always add the UDF to the metadata whether we found a duplicate or not.

Is there a test case that covers duplicates?
</blockquote></details>

Good catch, it should be a return. It's difficult to test directly whether the de-duplication is working (and I'm not sure it's worth it since it's only a performance issue), but I'll expand the test cases a bit.
___
*[`pkg/sql/sem/tree/function_name.go` line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NN6ENT9HLMi3njNbsZc-r2-68:-NNEPX7PC29E_NQ9fRYm:blz3aop) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2eaea52fe6d4f802bff5d8fadff770bd6971927/pkg/sql/sem/tree/function_name.go#L68)):*
>Something about this field name bothers me.

Changed it to `Reference` - does that make more sense to you?

> Also, I wonder if we should make this field unexported and have the act of resolving which rewrites this in place to do the swapping out into the field. Then have a method which uses the unexported field if it's populated and otherwise delegates to the underlying function reference.

I like this idea. Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
Don't forget to update the PR description with the new commit(s).

> Reviewed 17 of 17 files at r3, 4 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-:-NNc1XiaBjJ6Ox6HabPk:b-jbcesz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @DrewKimball, and @rharding6373)

___
*[`pkg/sql/delegate/show_function.go` line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NNc2-VoCunYneYdJLvr:-NNc2-VoCunYneYdJLvs:bgj58ya) ([raw file](https://github.com/cockroachdb/cockroach/blob/571685139fe7097bfd296b80a348f5e0f62f921f/pkg/sql/delegate/show_function.go#L42)):*
> ```Go
> 	var udfSchema string
> 	for _, o := range fn.Overloads {
> 		if o.HasSQLBody() {
> ```

nit: Can we leave this as `IsUDF`? It'd be odd to have `SHOW CREATE FUNCTION` work on just some built-ins.
___
*[`pkg/sql/opt/metadata.go` line 377 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NN9CfqO5f6Jt67vKKYX:-NNc3Gf27MSq7jOw9B8z:b-tjlp07) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2eaea52fe6d4f802bff5d8fadff770bd6971927/pkg/sql/opt/metadata.go#L377)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

We have to use name in case the name now resolves to a different overload - for example, in `logic_test/udf` the re-created sequence will return `5` instead of `1` if we don't properly handle this. It also wouldn't handle the database being switched.
</blockquote></details>

We talked IRL - there may be a way to use IDs instead of names.
___
*[`pkg/sql/opt/metadata.go` line 430 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NN9CzNe9z3XUzTlNDGs:-NNc4V_dAdT3m4maqCyF:b-jwnh1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2eaea52fe6d4f802bff5d8fadff770bd6971927/pkg/sql/opt/metadata.go#L430)):*
> That case is only when IsUDF is true, right? So it should be enough to check IsUDF || Body != "". I've added a method to Overload to handle this. If you think it's better to add HasBody, I'm fine with that as well.

Great idea - I like it.
___
*[`pkg/sql/opt/metadata.go` line 240 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NNc3ewVB1gd-6OBTwO8:-NNc3ewVB1gd-6OBTwO9:b3e78vv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b094c8e44c67faefda73eac52d3ebd052ec7734/pkg/sql/opt/metadata.go#L240)):*
> ```Go
> 
> 	if len(from.userDefinedFunctions) > 0 {
> 		for i := range from.userDefinedFunctions {
> ```

nit: `make` a new slice with the exact capacity if `md.userDefinedFunctions` is `nil`.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @mgartner, and @rharding6373)

___
*[`pkg/sql/delegate/show_function.go` line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NNc2-VoCunYneYdJLvr:-NOv1BO_5c5arhmqXIBS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/571685139fe7097bfd296b80a348f5e0f62f921f/pkg/sql/delegate/show_function.go#L42)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Can we leave this as `IsUDF`? It'd be odd to have `SHOW CREATE FUNCTION` work on just some built-ins.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/metadata.go` line 377 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NN9CfqO5f6Jt67vKKYX:-NOv1iUn8V6b-yGUGsPY:b-dbtbg1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2eaea52fe6d4f802bff5d8fadff770bd6971927/pkg/sql/opt/metadata.go#L377)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

We talked IRL - there may be a way to use IDs instead of names.
</blockquote></details>

I've refactored the dependency checking to also check the schemas that are used to resolve each object, as well as the current database if it was used during resolution. This allows us to check only IDs for all schema objects.
___
*[`pkg/sql/opt/metadata.go` line 240 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NNc3ewVB1gd-6OBTwO8:-NOv1hKzE8Yy4Cde92Z1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b094c8e44c67faefda73eac52d3ebd052ec7734/pkg/sql/opt/metadata.go#L240)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: `make` a new slice with the exact capacity if `md.userDefinedFunctions` is `nil`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
This should be ready for another look now.

DrewKimball commented:
>This should be ready for another look now.

Scratch that, looks like I have some more test failures to iron out.

DrewKimball committed:
opt: refactor metadata dependency tracking

This commit performs some refactoring for the way data source objects
are tracked in `opt.Metadata` in order to make future changes to UDT
and UDF dependency tracking easier. More particularly, UDTs and UDFs
will be able to re-resolve any references by name. This is necessary
in order to handle cases where changes to the search-path cause names
in the query to resolve to different objects.

Release note: None

DrewKimball committed:
opt: track UDT references by name in the Metadata

Previously, it was possible for invalid queries to be kept in the cache
after a schema change, since user-defined types were tracked only by OID.
This missed cases where the search path changed which object a name in
the query resolved to (or whether it resolved at all).

This patch fixes this behavior by tracking UDT references by name, similar
to what was already done for data sources. This ensures that the query
staleness check doesn't miss changes to the search path.

Fixes #96674

Release note (bug fix): Fixed a bug that could prevent a cached query with
a user-defined type reference from being invalidated even after a schema
change that should prevent the type from being resolved.

DrewKimball committed:
opt: track UDFs in the Metadata

This patch adds tracking for user-defined functions in `opt.Metadata`.
This ensures that the query cache will be correctly invalidated after
a schema change or search-path change that could change the query's
semantics, or cause it to error.

Fixes #93082
Fixes #93321

Release note (bug fix): Fixed a bug that could prevent a cached query
from being invalidated when a UDF referenced by that query was altered
or dropped.

DrewKimball commented:
There ended up being too many loose ends with the approach of always tracking objects by ID, so I've gone back to the original approach of keeping track of any names used to reference an object.

DrewKimball commented:
@mgartner Sorry for all the reviewable revisions, but this is passing all tests now.

mgartner reviewed (changes requested):
Great job working through all these tricky cases! It all looks great, just a few minor things to fixup.

nit: the commits in the PR desriptions are in descending order. It's easier to follow if they are ascending. If you add this to your `~/.gitconfig`:

```
[alias]
  pr-notes = "!git --no-pager log upstream/master..HEAD --reverse --pretty='format:#### %s%n%n%b'"
```

Then you can auto-generate the PR notes with `git pr-notes`.

> Reviewed 2 of 15 files at r5, 1 of 25 files at r9, 1 of 9 files at r12, 26 of 26 files at r17, 11 of 11 files at r21, 5 of 5 files at r22, 9 of 9 files at r23, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-:-NQ24XH35tb1wcbXkxrW:b-ulm0up)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @DrewKimball, and @rharding6373)

___
*[`pkg/sql/opt/metadata.go` line 261 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NQ25wPwEcrvicrrknjO:-NQ25wPwEcrvicrrknjP:b-1w86zs) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee4f3ef2e46a19a34ff9a9adbf18d02ca9b40ce5/pkg/sql/opt/metadata.go#L261)):*
> ```Go
> 	}
> 
> 	for id, dataSource := range from.dataSourceDeps {
> ```

These three loops all do the same thing. Are we missing a test that might have caught this?
___
*[`pkg/sql/opt/metadata.go` line 343 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NQ29zMM-B87tfn00sxR:-NQ29zMM-B87tfn00sxS:b-s5btrk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee4f3ef2e46a19a34ff9a9adbf18d02ca9b40ce5/pkg/sql/opt/metadata.go#L343)):*
> ```Go
> 			}
> 		} else {
> 			toCheck, _, err = optCatalog.ResolveDataSourceByID(ctx, cat.Flags{}, dataSource.ID())
> ```

Can you add a comment explaining this `else` case more? I'm assuming it's necessary when a query references a datasource by ID. Is that correct?
___
*[`pkg/sql/logictest/testdata/logic_test/udf` line 3237 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NQ2CBCf99jQPauzrMiD:-NQ2CBCf99jQPauzrMiE:bd7kdzo) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc8b79274274330852270c7fcf0bdc4e09ebdde/pkg/sql/logictest/testdata/logic_test/udf#L3237)):*
> ```
> SELECT abs(1);
> ----
> 1
> ```

This should be 101, not 1, correct?


<!-- Sent from Reviewable.io -->

DrewKimball commented:
>Then you can auto-generate the PR notes with git pr-notes.

TIL, thanks!

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @mgartner, and @rharding6373)

___
*[`pkg/sql/opt/metadata.go` line 261 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NQ25wPwEcrvicrrknjO:-NQ9SjOlDZKlsO8PCSJd:bfak9si) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee4f3ef2e46a19a34ff9a9adbf18d02ca9b40ce5/pkg/sql/opt/metadata.go#L261)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

These three loops all do the same thing. Are we missing a test that might have caught this?
</blockquote></details>

Yeesh, good catch. Added some tests.
___
*[`pkg/sql/opt/metadata.go` line 343 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NQ29zMM-B87tfn00sxR:-NQ9bRjs8jVGpKx6KJLk:badvdv8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee4f3ef2e46a19a34ff9a9adbf18d02ca9b40ce5/pkg/sql/opt/metadata.go#L343)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Can you add a comment explaining this `else` case more? I'm assuming it's necessary when a query references a datasource by ID. Is that correct?
</blockquote></details>

That's right. Added a comment.
___
*[`pkg/sql/logictest/testdata/logic_test/udf` line 3237 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NQ2CBCf99jQPauzrMiD:-NQ9eRW8A-tk77Q7w5A3:b-juggcr) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc8b79274274330852270c7fcf0bdc4e09ebdde/pkg/sql/logictest/testdata/logic_test/udf#L3237)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This should be 101, not 1, correct?
</blockquote></details>

Hm... another good catch. Turns out we also need to track builtin functions to handle user-defined overrides.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!

> Reviewed 12 of 12 files at r24, 4 of 4 files at r25, 5 of 5 files at r26, 9 of 9 files at r27, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-:-NQCDkO93YSVhjpbHZfz:bcmw7ss)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, and @rharding6373)

___
*[`pkg/sql/logictest/testdata/logic_test/udf` line 3237 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-NQ2CBCf99jQPauzrMiD:-NQCDYPwEXeEZGowzv2M:bnwwqe3) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc8b79274274330852270c7fcf0bdc4e09ebdde/pkg/sql/logictest/testdata/logic_test/udf#L3237)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Hm... another good catch. Turns out we also need to track builtin functions to handle user-defined overrides.
</blockquote></details>

:+1:  Feel free to merge this now and address in a follow-up PR. You can omit these tests, or leave them and leave a TODO pointing out the incorrect result, up to you.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
>👍 Feel free to merge this now and address in a follow-up PR. You can omit these tests, or leave them and leave a TODO pointing out the incorrect result, up to you.

Removed the tests for now, and I'll write up a fix soon since it seems straightforward on the face of it.

DrewKimball commented:
TFYRs!

DrewKimball commented:
Test failure is a known flake: https://github.com/cockroachdb/cockroach/issues/98404

bors r+

mgartner commented:
bors r+

mgartner commented:
@rharding6373 can you re-review when you get a chance? Looks like bors now requires there to be no changes requested from reviewers.

mgartner commented:
@DrewKimball looks like you need to rebase.

@rharding6373 The merge is blocked until you approve.

rharding6373 reviewed (approved):
Sorry about the delay.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96045#-:-NQx61qh6RHg7g7XCeZC:bvb4vdh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @chengxiong-ruan, and @mgartner)



<!-- Sent from Reviewable.io -->

DrewKimball commented:
TFYRs!

bors r+
